### DISCUSSION TITLE
Why does ESLint's RuleTester no longer work with @babel/eslint-parser >= 7.14.2?

### DISCUSSION THREAD

**Discussion by stefcameron** [babel/babel#13391] [Q&A]
As of 7.14.3, I get this error when trying to use the RuleTester export of `eslint`:

```
Cannot find module '/src/project/node_modules/@babel/eslint-parser/lib' from 'node_modules/eslint/lib/rule-tester/rule-tester.js'
```

I'm guessing it's maybe something to do with the fact that `/src/project/node_modules/@babel/eslint-parser/lib` now contains an `index.cjs` file instead of `index.js`.

I'm running Node 14.16.1 on macOS 11.3.1.

I'm thinking something somewhere isn't considering `index.cjs` as a default file to load for a module definition at `/src/project/node_modules/@babel/eslint-parser/lib` because in `@babel/eslint-parser@7.14.2`, the files are different, we find `index.js` in there.

What am I supposed to do to get this to work now?

**Comment by nicolo-ribaudo** [comment]
I'm surprised that it doesn't work since we specify the package entry point (https://github.com/babel/babel/blob/813f2a3cce04a5366de7c1cc370652db64507305/eslint/babel-eslint-parser/package.json#L22). How are you using `RuleTester`?

**Reply by stefcameron** [reply]
Just like this:
```
const rules = require('../../../eslint-local-rules');
const { RuleTester } = require('eslint');

const ruleTester = new RuleTester();
```

I think the problem (based on the error) is that for some reason, it's `node_modules/@babel/eslint-parser/lib` that's specifically referenced, not `node_modules/@babel/eslint-parser` which has a `package.json` with `main` that points to `./lib/index.cjs` as you pointed out.

I don't know why it's doing that, but that's the only thing I can think of so far. Something related to `RuleTester` tries to do a load directly in `node_modules/@babel/eslint-parser/lib` and because `index.js` doesn't exist there anymore, Node just doesn't load anything.

**Reply by nicolo-ribaudo** [reply]
We can add an `index.js` file back for backward compatiblity (even if the "public interface" should just be the `@babel/eslint-parser` entrypoint), although I'm surprised that `RuleTester` adds `/lib`. How do you enable `@babel/eslint-parser` (by default ESLint uses its own parser)?

**Reply by stefcameron** [reply]
That would be good! I tried renaming `index.cjs` to `index.js` in the `lib` directory, and it worked. I also tried adding a new `package.json` file in `lib` and specified `"main": "index.cjs"` and that worked as well. So 2 options for you! 🙂 

All I do is specify ` parser: '@babel/eslint-parser',` in my `.eslintrc.js` file that exports my ESLint configuration. Nothing else.

**Reply by nicolo-ribaudo** [reply]
I opened https://github.com/babel/babel/pull/13397. I still don't understand why this happens; what ESLint version are you using?

**Reply by JLHwung** [reply]
https://github.com/eslint/eslint/blob/2e43dacd24337a82d4184fac9b44d497675f46ef/lib/rule-tester/rule-tester.js#L523-L529

The `config.parser` in rule tester only accepts absolute path of parser, I guess the `config.parser` is passed as `path.dirname(require.resolve("eslint-parser"))` which unfortunately bypasses the `exports` entry isolation.

**Reply by nicolo-ribaudo** [reply]
It seems like ESLint only does `require.resolve` (https://github.com/eslint/eslintrc/blob/7c43d7784e39cf0b7b102af64f703cade11252bb/lib/config-array-factory.js#L929, https://github.com/eslint/eslintrc/blob/7c43d7784e39cf0b7b102af64f703cade11252bb/lib/shared/relative-module-resolver.js#L28), and then just `require()`s it (https://github.com/eslint/eslintrc/blob/7c43d7784e39cf0b7b102af64f703cade11252bb/lib/config-array-factory.js#L934).

@stefcameron Could you try upgrading to the last ESLint version, just to make sure that it wasn't a bug in an old ESLint version that ignored the `main` field and/or used `path.dirname`?

**Reply by stefcameron** [reply]
@nicolo-ribaudo I was already testing with the latest version of ESLint `7.27.0`. I've been on that for a while now. I saw this issue with `@babel/eslint-parser` version `7.14.2` and every version since.

**Reply by JLHwung** [reply]
I can not reproduce this issue on the Babel repo: We use `eslint.RuleTester` and imports `@babel/eslint-parser`, too:

https://github.com/babel/babel/pull/13420/files#diff-880066e4af576531b0519e916adb5dd1fcbfb622237e082827ee2acc47d72734R4

Can you provide a reproducing repo?

**Reply by stefcameron** [reply]
@JLHwung Thank you for that link to some code. Turns out we still had code that was referencing `@babel/eslint-parser/lib` directly, which must have been necessary prior to 7.14.2 or perhaps a while back, and continued working until the `lib/index.js` was renamed `lib/index.cjs` (in a patch!) and that no longer worked. Referencing `@babel/eslint-parser` for our RuleTester parser configuration, as you pointed to there in that commit solved our issue.

**Reply by JLHwung** [reply]
`@babel/eslint-parser/lib` is never supported as it depends on internal file structures. Node >= 14 will throw on `require("@babel/parser/lib")` because `/lib` is not exported in `packageJson.exports`.

**Reply by stefcameron** [reply]
But I've been running this on Node 14 (whatever the latest LTS is, 14.17.0 now) for quite a while as well, no exceptions thrown with that code. The only thing that gave me trouble here is the renaming of `index.js` to `index.cjs`. There had to have been a time when this `/lib` reference was necessary. Perhaps it hasn't been necessary for a while, but it kept working until there was no longer a default `index.js` file in `/lib`. 🤷‍♂️

**Reply by stefcameron** [reply]
But I do recognize `/lib` is an internal part of the module, of course, and I'm glad to see it's no longer needed. I don't do this type of reference on a normal basis. :)
